E2_ESRGAN/lib/dataset.py

```diff
@@ -1,44 +1,105 @@
+
     low_resolution = tf.image.resize(
-        image, [dimension // factor, dimension // factor], method=method)
-    return (low_resolution, high_resolution)
+        high_resolution,
+        [dimension // factor, dimension // factor],
+        method=method)
+    low_resolution = tf.clip_by_value(low_resolution, 0, 255)
+    high_resolution = tf.clip_by_value(high_resolution, 0, 255)
+    return low_resolution, high_resolution
+  scale_fn.dimension = dimension
   return scale_fn
 
 
+def augment_image(
+        low_res_map_fn,
+        brightness_delta=0.05,
+        contrast_factor=[0.7, 1.3],
+        saturation=[0.6, 1.6]):
+  
+  def augment_fn(_, high_resolution, *args, **kwargs):
+    high_resolution = tf.image.random_flip_left_right(high_resolution)
+    high_resolution = tf.image.random_flip_up_down(high_resolution)
+    high_resolution = tf.image.random_brightness(
+        high_resolution, max_delta=brightness_delta)
+    if contrast_factor:
+      high_resolution = tf.image.random_contrast(
+          high_resolution, *contrast_factor)
+    if saturation:
+      high_resolution = tf.image.random_saturation(
+          high_resolution, *saturation)
+    low_res, high_res = low_res_map_fn(high_resolution)
+    return low_res, high_res
+  return augment_fn
+
+
+def reform_dataset(dataset, types):
+  """ Helper function to convert the output_dtype of the dataset
+      from (tf.float32, tf.uint8) to desired dtype
+  """
+  def generator_fn():
+    for data in dataset:
+      yield data[0], data[1]
+  return tf.data.Dataset.from_generator(
+      generator_fn, types, (tf.TensorShape([None, None, 3]), tf.TensorShape(None)))
+
+
 def load_dataset_directory(
         name,
         directory,
         low_res_map_fn,
         batch_size=32,
         iterations=1,
         shuffle=True,
+        augment=True,
         cache_dir="cache/",
         buffer_size=io.DEFAULT_BUFFER_SIZE):
 
   if not tf.io.gfile.exists(cache_dir):
     tf.io.gfile.mkdir(cache_dir)
   dl_config = tfds.download.DownloadConfig(manual_dir=directory)
-  dataset = (tfds.load("image_label_folder/dataset_name=%s" % name,
-                       split="train",
-                       as_supervised=True,
-                       download_and_prepare_kwargs={"download_config": dl_config})
-             .map(low_res_map_fn, num_parallel_calls=tf.data.experimental.AUTOTUNE)
-             .batch(batch_size)
-             .prefetch(buffer_size)
-             .cache(cache_dir))
+
+  dataset = reform_dataset(
+      tfds.load(
+          "image_label_folder/dataset_name=%s" %
+          name,
+          split="train",
+          as_supervised=True,
+          download_and_prepare_kwargs={
+              "download_config": dl_config}),
+      (tf.float32, tf.float32))
+
+  dataset = (dataset.filter(
+      lambda image, *args: tf.greater_equal(
+          image.shape[:-1],
+          2 * [low_res_map_fn.dimension]).numpy().all())
```

Comment:
> well, if I assign this dimension, then I can fetch this attribute in augment_fn, else, I need to pass the dimension again. It's just a way to pass the same param again and again, into functions which will be actually chain mapped on the dataset.

@srjoglekar246 , this the dimension parameter I mentioned. Instead of passing an extra parameter, I used this hack to get the dimension from the `low_res_map_fn`.